src/System.Private.CoreLib/shared/System/Runtime/CompilerServices/AsyncMethodBuilder.cs

```diff
@@ -501,21 +492,55 @@ public void SetStateMachine(IAsyncStateMachine stateMachine)
             // cases is we lose the ability to properly step in the debugger, as the debugger uses that
             // object's identity to track this specific builder/state machine.  As such, we proceed to
             // overwrite whatever's there anyway, even if it's non-null.
-            var box = new AsyncStateMachineBox<TStateMachine>();
+            var box = AsyncMethodBuilderCore.TrackAsyncMethodCompletion ?
+                CreateDebugFinalizableAsyncStateMachineBox<TStateMachine>() :
+                new AsyncStateMachineBox<TStateMachine>();
             m_task = box; // important: this must be done before storing stateMachine into box.StateMachine!
             box.StateMachine = stateMachine;
             box.Context = currentContext;
             return box;
         }
 
+        // Avoid forcing the JIT to build DebugFinalizableAsyncStateMachineBox<TStateMachine> unless it's actually needed.
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private static AsyncStateMachineBox<TStateMachine> CreateDebugFinalizableAsyncStateMachineBox<TStateMachine>()
+            where TStateMachine : IAsyncStateMachine =>
+            new DebugFinalizableAsyncStateMachineBox<TStateMachine>();
+
+        /// <summary>
+        /// Provides an async state machine box with a finalizer that will fire an EventSource
+        /// event about the state machine if it's being finalized without having been completed.
+        /// </summary>
+        /// <typeparam name="TStateMachine">Specifies the type of the state machine.</typeparam>
+        private sealed class DebugFinalizableAsyncStateMachineBox<TStateMachine> : // SOS DumpAsync command depends on this name
```

Comment:
I would ifdef out `CreateDebugFinalizableAsyncStateMachineBox` and `DebugFinalizableAsyncStateMachineBox` for CoreRT. 

DebugFinalizableAsyncStateMachineBox looks like a small type, but it actually is not because of it will have copy of all the slots from its parent. It will add another hundred(s) bytes per each async method in CoreRT / ProjectN binaries without adding much value.

There is a different better way to implement this debug-aid that is AOT / code footprint friendly: Have a non-generic tracker type with finalizer that is attached to the AsyncStateMachine via ConditionalWeakHandle.

Reply:
I placed `#if CORERT` around the code and removed the changes to `TplEtwProvider`. I like the implementation you propose, but I'm not planning to implement it now.